hatch/src/bin/hatch.rs

```diff
@@ -1,47 +1,64 @@
+extern crate clap;
 extern crate hatch;
+extern crate yaml_rust;
 
 use std::collections::HashMap;
 
-use hatch::cli::Cli;
-use hatch::cli::commands::Command;
-use hatch::cli::commands::new::New;
-use hatch::cli::commands::update::Update;
+use clap::App;
 use hatch::cli::commands::build::Build;
-use hatch::cli::commands::test::Test;
+use hatch::cli::commands::new::New;
 use hatch::cli::commands::run::Run;
+use hatch::cli::commands::test::Test;
+use hatch::cli::commands::update::Update;
+use hatch::cli::commands::Command;
+use hatch::constants;
+use hatch::hatch_error::{HatchResult, MissingParameterError};
+use yaml_rust::YamlLoader;
+
+fn do_me_a_hatch() -> HatchResult<()> {
+    let cli = include_str!("cli.yml");
+    let docs = YamlLoader::load_from_str(cli)?;
+    let doc = &docs[0];
+    let matches = App::from_yaml(doc).get_matches();
+    let (name, args) = matches.subcommand();
+
+    // TODO: get the generator from the CLI and inject into commands
+    // let generator;
 
-fn main() {
-    // create the subcommand to command map
     let mut subcommands: HashMap<&'static str, Box<Command>> = HashMap::new();
+    let run_command = Box::new(Run::new());
+    subcommands.insert(constants::RUN_NAME, run_command);
 
     let new_command = Box::new(New::new());
-    subcommands.insert(new_command.subcommand_name(), new_command);
+    subcommands.insert(constants::NEW_NAME, new_command);
 
     let update_command = Box::new(Update::new());
-    subcommands.insert(update_command.subcommand_name(), update_command);
+    subcommands.insert(constants::UPDATE_NAME, update_command);
 
     let build_command = Box::new(Build::new());
-    subcommands.insert(build_command.subcommand_name(), build_command);
+    subcommands.insert(constants::BUILD_NAME, build_command);
 
     let test_command = Box::new(Test::new());
-    subcommands.insert(test_command.subcommand_name(), test_command);
+    subcommands.insert(constants::TEST_NAME, test_command);
 
-    let run_command = Box::new(Run::new());
-    subcommands.insert(run_command.subcommand_name(), run_command);
-
-    // initialize cli with the set of subcommand
-    let cli = Cli::new(
-        subcommands
-            .values()
-            .map(|v| v.cli_subcommand())
-            .collect::<Vec<_>>(),
-    );
-
-    // execute selected subcommand
-    let (name, args) = cli.subcommand();
-    if let Some(cmd) = subcommands.get(name) {
-        if let Err(e) = cmd.execute(args.unwrap()) {
-            println!("{}", e);
+    let subcommand = subcommands.get(name).ok_or_else(|| MissingParameterError)?;
+
+    // TODO: args.unwrap is bad, don't do this, it should be resultified
```

Comment:
I believe that we decided that unwrap is ok in this case?

Reply:
yeah we probably did, let's remove this comment

Reply:
Agreed.